subsys/bluetooth/host/l2cap.c

```diff
@@ -903,63 +1038,200 @@ static void le_conn_req(struct bt_l2cap *l2cap, u8_t ident,
+	if (!server) {
+		result = BT_L2CAP_LE_ERR_PSM_NOT_SUPP;
+		goto response;
+	}
 
-		/* Set channel PSM */
-		chan->psm = server->psm;
+	/* Check if connection has minimum required security level */
+	if (conn->sec_level < server->sec_level) {
+		result = BT_L2CAP_LE_ERR_AUTHENTICATION;
+		goto response;
+	}
 
-		/* Update state */
-		bt_l2cap_chan_set_state(chan, BT_L2CAP_CONNECTED);
+	while (buf->len >= sizeof(scid)) {
+		scid = net_buf_pull_le16(buf);
 
-		if (chan->ops->connected) {
-			chan->ops->connected(chan);
+		result = l2cap_chan_accept(conn, server, scid, mtu, mps,
+					   credits, &chan[i]);
+		switch (result) {
+		case BT_L2CAP_LE_SUCCESS:
+			ch = BT_L2CAP_LE_CHAN(chan[i]);
+			dcid[i++] = sys_cpu_to_le16(ch->rx.cid);
+			continue;
+		/* Some connections refused – invalid Source CID */
+		case BT_L2CAP_LE_ERR_INVALID_SCID:
+		/* Some connections refused – Source CID already allocated */
+		case BT_L2CAP_LE_ERR_SCID_IN_USE:
+			/* If a Destination CID is 0x0000, the channel was not
+			 * established.
+			 */
+			dcid[i++] = 0x0000;
+			continue;
+		/* Some connections refused – not enough resources
+		 * available.
+		 */
+		case BT_L2CAP_LE_ERR_NO_RESOURCES:
+		default:
+			goto response;
 		}
+	}
 
-		/* Prepare response protocol data */
-		rsp->dcid = sys_cpu_to_le16(ch->rx.cid);
+response:
+	if (!i) {
+		i = buf->len / sizeof(scid);
+	}
+
+	buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
+				      sizeof(*rsp) + (sizeof(scid) * i));
+
+	rsp = net_buf_add(buf, sizeof(*rsp));
+	(void)memset(rsp, 0, sizeof(*rsp));
+
+	if (result == BT_L2CAP_LE_ERR_UNACCEPT_PARAMS ||
+	    result == BT_L2CAP_LE_ERR_PSM_NOT_SUPP ||
+	    result == BT_L2CAP_LE_ERR_AUTHENTICATION) {
+		memset(dcid, 0, sizeof(scid) * i);
+	} else if (ch) {
 		rsp->mps = sys_cpu_to_le16(ch->rx.mps);
 		rsp->mtu = sys_cpu_to_le16(ch->rx.mtu);
 		rsp->credits = sys_cpu_to_le16(ch->rx.init_credits);
-		rsp->result = BT_L2CAP_LE_SUCCESS;
-	} else {
-		rsp->result = sys_cpu_to_le16(BT_L2CAP_LE_ERR_NO_RESOURCES);
 	}
-rsp:
+
+	net_buf_add_mem(buf, dcid, sizeof(scid) * i);
+
+	rsp->result = sys_cpu_to_le16(result);
+
+	bt_l2cap_send(conn, BT_L2CAP_CID_LE_SIG, buf);
+}
+
+static void le_ecred_reconf_req(struct bt_l2cap *l2cap, u8_t ident,
+				struct net_buf *buf)
+{
+	struct bt_conn *conn = l2cap->chan.chan.conn;
+	struct bt_l2cap_ecred_reconf_req *req;
+	struct bt_l2cap_ecred_reconf_rsp *rsp;
+	u16_t mtu, mps;
+	u16_t scid, result;
+
+	if (buf->len < sizeof(*req)) {
+		BT_ERR("Too small ecred reconf req packet size");
+		return;
+	}
+
+	req = net_buf_pull_mem(buf, sizeof(*req));
+
+	mtu = sys_le16_to_cpu(req->mtu);
+	mps = sys_le16_to_cpu(req->mps);
+
+	if (mtu < L2CAP_ECRED_MIN_MTU) {
+		result = BT_L2CAP_RECONF_INVALID_MTU;
+		goto response;
+	}
+
+	if (mps < L2CAP_ECRED_MIN_MTU) {
+		result = BT_L2CAP_RECONF_INVALID_MTU;
+		goto response;
+	}
+
+	BT_DBG("mtu %u mps %u", mtu, mps);
+
+	result = BT_L2CAP_RECONF_SUCCESS;
```

Comment:
You might want to move the this after the loop, since you're needlessly writing twice to that variable in the invalid MTU case.

Reply:
Done.